cmd/webhook/main.go

```diff
@@ -37,58 +43,96 @@ const (
+	v1alpha1.SchemeGroupVersion.WithKind("EventListener"):         &v1alpha1.EventListener{},
+	v1alpha1.SchemeGroupVersion.WithKind("TriggerBinding"):        &v1alpha1.TriggerBinding{},
+	v1alpha1.SchemeGroupVersion.WithKind("TriggerTemplate"):       &v1alpha1.TriggerTemplate{},
+}
 
-	clusterConfig, err := rest.InClusterConfig()
-	if err != nil {
-		log.Fatalf("Failed to get in cluster config: %v", err)
-	}
+func NewDefaultingAdmissionController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
 
-	kubeClient, err := kubernetes.NewForConfig(clusterConfig)
-	if err != nil {
-		log.Fatalf("Failed to get the client set: %v", err)
-	}
+	store := defaultconfig.NewStore(logging.FromContext(ctx).Named("config-store"))
+	store.WatchConfigs(cmw)
 
-	logger := logging.ConfigureLogging(WebhookLogKey, ConfigName, stopCh, kubeClient)
-
-	defer func() {
-		err := logger.Sync()
-		if err != nil {
-			logger.Fatalf("Failed to sync the logger", zap.Error(err))
-		}
-	}()
-
-	options := webhook.ControllerOptions{
-		ServiceName:                     "tekton-triggers-webhook",
-		DeploymentName:                  "tekton-triggers-webhook",
-		Namespace:                       system.GetNamespace(),
-		Port:                            8443,
-		SecretName:                      "triggers-webhook-certs",
-		WebhookName:                     "triggers-webhook.tekton.dev",
-		ResourceAdmissionControllerPath: "/",
-	}
-	resourceHandlers := map[schema.GroupVersionKind]webhook.GenericCRD{
-		v1alpha1.SchemeGroupVersion.WithKind("ClusterTriggerBinding"): &v1alpha1.ClusterTriggerBinding{},
-		v1alpha1.SchemeGroupVersion.WithKind("EventListener"):         &v1alpha1.EventListener{},
-		v1alpha1.SchemeGroupVersion.WithKind("TriggerBinding"):        &v1alpha1.TriggerBinding{},
-		v1alpha1.SchemeGroupVersion.WithKind("TriggerTemplate"):       &v1alpha1.TriggerTemplate{},
-	}
-	resourceAdmissionController := webhook.NewResourceAdmissionController(resourceHandlers, options, true)
-	admissionControllers := map[string]webhook.AdmissionController{
-		options.ResourceAdmissionControllerPath: resourceAdmissionController,
-	}
+	return defaulting.NewAdmissionController(ctx,
 
-	// Decorate contexts with the current state of the config.
-	ctxFunc := v1alpha1.WithUpgradeViaDefaulting
+		// Name of the resource webhook.
+		"webhook.triggers.tekton.dev",
 
-	controller, err := webhook.New(kubeClient, options, admissionControllers, logger, ctxFunc)
-	if err != nil {
-		logger.Fatal("Error creating admission controller", zap.Error(err))
-	}
+		// The path on which to serve the webhook.
+		"/defaulting",
+
+		// The resources to validate and default.
+		types,
+
+		// A function that infuses the context passed to Validate/SetDefaults with custom metadata.
+		v1alpha1.WithUpgradeViaDefaulting,
+		//func(ctx context.Context) context.Context {
+		//	//return v1alpha1.WithUpgradeViaDefaulting(ctx)
+		//	return contexts.WithUpgradeViaDefaulting(store.ToContext(ctx))
+		//},
```

Comment:
You can cut this :stuck_out_tongue_closed_eyes:

Reply:
:face_with_head_bandage: